Implement custom provider record announcements for the DSN (v2).

shamil-gadelshin opened the pull request:
This PR follows comments from the previous version node announcement PR (https://github.com/subspace/subspace/pull/1430) Fixes https://github.com/subspace/subspace/issues/1188

Previously, we migrated to our custom fork of rust-libp2p because we needed more control over Kademlia provider announcements. It worked for a while but keeping our fork updated is not very convenient, additionally, we wanted to have more control over the announcements because the current performance isn't great.

This PR implements custom provider records announcements.

### Main changes
- some refactoring related to piece index and pieces.rs module
- new request-response handler
- new announcement algorithm
- examples update
- exporting of external addresses to populate provider records
- removing Kademlia announcement calls and handlers

The following PR will update rust-libp2p fork.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

shamil-gadelshin committed:
Prepare networking for custom announcements.

- remove Kademlia announcements
- prepare new announcement logic
- export external addresses to fill ProviderRecord
- change networking examples

shamil-gadelshin committed:
Add custom announcements.

- add new request type
- add farmer and node request handler

nazar-pc commented on crates/subspace-farmer/src/utils/farmer_provider_storage.rs:
```diff
@@ -107,7 +109,7 @@ where
                 provider: self.local_peer_id,
                 expires: None,
-                addresses: Vec::new(), // TODO: add address hints
+                addresses: Vec::new(), // Kademlia adds addresses for local providers
```
I do not understand what this comment adds. Nothing adds anything on this line 🤷

shamil-gadelshin replied:
Here is the Kademlia implementation code for reference: https://github.com/libp2p/rust-libp2p/blob/3c5940aeadb9ed8527b6f7aa158797359085293d/protocols/kad/src/behaviour.rs#L911

nazar-pc replied:
Comment in libp2p makes perfect sense, now I understand what you meant, but you better copy the whole comment from libp2p or re-phrase it, indicating that Kademlia will add it later for locally provided records. The way it is written right now is confusing.

Which in turn doesn't make sense anymore because we have supposedly removed Kademlia's implementation 🤷

shamil-gadelshin replied:
I copied the comment. We removed Kademlia implementation only for announcements because we had issues with it. `get_providers` operation works well and it will be suboptimal to replace it with `get_closest_peers` + custom requests. Please, let me know if I understood you incorrectly.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm/dsn.rs:
```diff
@@ -371,3 +438,19 @@ pub(crate) fn start_announcements_processor(
+    pub(crate) fn announce(&self, provider_record: &ProviderRecord) {
+        self.node.announce(provider_record);
+    }
+}
```
This data structure doesn't make much sense to me, why is it necessary? It seems like we could just have a single-line method call instead.

shamil-gadelshin replied:
I moved the code from `Node` here. Hope it makes more sense now.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm/dsn.rs:
```diff
@@ -156,69 +160,127 @@ pub(super) fn configure_dsn(
 
-    let default_config = Config::new(protocol_prefix, keypair, farmer_provider_storage);
+    let provider_record_announcer: Arc<RwLock<Option<NodeProviderRecordAnnouncer>>> =
+        Arc::new(RwLock::new(None));
```
Are you sure this doesn't cause circular data structure leaking memory? Looks like what you wanted here is to store weak reference to `Node` and upgrade it right before use, possibly with https://docs.rs/arc-swap

nazar-pc commented on crates/subspace-networking/src/request_handlers/piece_announcement.rs:
```diff
@@ -0,0 +1,52 @@
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub struct PieceAnnouncementRequest {
+    /// Request key - piece index multihash
+    pub piece_key: Vec<u8>,
```
We did not have `piece_key` anywhere in codebase before. Should this be `piece_index_hash: Multihash` maybe?

shamil-gadelshin replied:
We used `piece_index_hash` previously for the actual piece index hash, but I don't mind renaming it here.

nazar-pc commented on crates/subspace-networking/src/request_handlers/piece_announcement.rs:
```diff
@@ -0,0 +1,52 @@
+    pub piece_key: Vec<u8>,
+
+    /// External addresses of the peer
+    pub addresses: Vec<Vec<u8>>,
```
Why are they in raw bytes? Looks like having `Multiaddr` here would make much more sense and make somewhat confusing `converted_addresses` unnecessary.

shamil-gadelshin replied:
`scale` codec doesn't support libp2p types.

nazar-pc replied:
That only means you can't derive implementation, but you can still implement it manually, `Multiaddr` is not hard to convert to bytes and back. You are doing it anyway, the only difference is that instead of doing it in one place and work with mostly stack-allocated data you're doing it in every place where `PieceAnnouncementRequest` is used and have to do multiple allocations in `Vec<Vec<u8>>`.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm/dsn.rs:
```diff
@@ -156,69 +160,127 @@ pub(super) fn configure_dsn(
+                            provider_record_announcer.announce(&provider_record);
+                        }
+
+                        Some(PieceAnnouncementResponse)
```
I would have expected this to be the same as `node.on_announcement`, why is it completely different?

nazar-pc commented on crates/subspace-networking/src/request_handlers/piece_announcement.rs:
```diff
@@ -0,0 +1,52 @@
+
+/// Piece announcement protocol response.
+#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
+pub struct PieceAnnouncementResponse; // just an acknowledgement
```
Would it maybe make sense to have an enum with one variant, is there a chance we might want to add another enum variant here in the future?

nazar-pc reviewed (commented):
After reading both commits I think you wanted to keep both old and new announcement mechanisms. I'm not entirely sure why though since we already broke compatibility with Gemini 3d and even then I'm not entirely sure why would any refactoring be necessary there.

There are comments answering "what?", but not "why?", I'm left guessing what you wanted to do, which options considered and why decided to go with this particular approach. It would be helpful to have that information before review, so that it is easier to map changes onto the desired outcome. It doesn't have to be a detailed essay, but at least something.

shamil-gadelshin committed:
Refactor record announcements.

- also apply review suggestions: make announcement response enum, rename fields of the announcement request

shamil-gadelshin reviewed (commented):
> After reading both commits I think you wanted to keep both old and new announcement mechanisms. I'm not entirely sure why though since we already broke compatibility with Gemini 3d and even then I'm not entirely sure why would any refactoring be necessary there.
> There are comments answering "what?", but not "why?", I'm left guessing what you wanted to do, which options considered and why decided to go with this particular approach. It would be helpful to have that information before review, so that it is easier to map changes onto the desired outcome. It doesn't have to be a detailed essay, but at least something.

Actually, I removed the old mechanism. The only leftover was handler storage in the node which caused a lot of unclear code. I simplified it. Please, have another look - I hope it makes more sense now. The implementation is straightforward and according to our discussion - basically, it's `get_closest_peers` + manual announcement requests. Please, let me know if it is still not clear - I'll provide more context.

shamil-gadelshin committed:
Merge branch 'main' into node-announcements3

nazar-pc commented on crates/subspace-networking/src/request_handlers/piece_announcement.rs:
```diff
@@ -44,7 +44,9 @@ impl GenericRequest for PieceAnnouncementRequest {
 #[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
-pub struct PieceAnnouncementResponse; // just an acknowledgement
+pub enum PieceAnnouncementResponse {
+    Success, // just an acknowledgement
```
Come on, this is Rust:
```suggestion
    /// just an acknowledgement
    Success,
```

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm/dsn.rs:
```diff
@@ -381,11 +372,12 @@ pub(crate) fn start_announcements_processor(
         mpsc::channel(MAX_CONCURRENT_ANNOUNCEMENTS_QUEUE.get());
 
-    let handler_id = node.on_announcement(Arc::new({
+    let handler_id = provider_record_announcer.on_announcement(Arc::new({
```
Now I'm wondering why we need `ProviderRecordAnnouncer`. Comment of this function says that it handles announcements, which was true before, but now it is a lie. Actual processing is now split between this function and `PieceAnnouncementRequestHandler::create`'s callback. Why not make it in the same place such that you don't have to jump through additional abstractions like `ProviderRecordAnnouncer` to understand what is going on here.

shamil-gadelshin replied:
The confusion here seems originated in the naming and comments. We need to pass the newly received records to the records processor (download piece, save to local piece, reannounce) and this object handles the re-announcement. This is why this functionality is situated only in the farmer rather than in the networking crate or in both farmer and the node. I will change the name and the comment to clarify its purpose.

shamil-gadelshin committed:
Add scale encoding for piece announcement request.

- add more comments
- rename RecordProcessorAnnouncementHelper

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm/dsn.rs:
```diff
@@ -314,23 +357,21 @@ pub(crate) fn start_announcements_processor(
         mpsc::channel(MAX_CONCURRENT_ANNOUNCEMENTS_QUEUE.get());
 
-    let handler_id = node.on_announcement(Arc::new({
+    let handler_id = record_processor_helper.on_announcement(Arc::new({
```
What if we remove `record_processor_helper` and instead have `provider_records_receiver` as an argument of this function and sender can be called from `PieceAnnouncementRequestHandler::create`? I think it might be easier to follow that way.

shamil-gadelshin replied:
I'm not sure I understood 100% your proposal. The latest commit is the best effort to address your suggestion.

nazar-pc replied:
My point was that we don't need these abstractions. We can remove both `RecordProcessorAnnouncementHelper` and related type aliases.

nazar-pc reviewed (commented):
I didn't expect initially that we'd still use libp2p's providers for queries. Considering you decided to still use it (I do think we'll have to get rid of it as well) it is fine. There are some merge conflicts though.

shamil-gadelshin committed:
Refactor farmer record processor.

shamil-gadelshin committed:
Merge branch 'main' into node-announcements3

# Conflicts:
#	crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs
#	crates/subspace-networking/src/node_runner.rs

shamil-gadelshin committed:
Remove RecordProcessorAnnouncementHelper

nazar-pc committed:
Prune more abstractions

nazar-pc committed:
Adjust `ProviderStorage` interface enabling code simplifications

nazar-pc committed:
Merge remote-tracking branch 'origin/main' into node-announcements3

# Conflicts:
#	crates/subspace-farmer/src/bin/subspace-farmer/commands/farm/dsn.rs
